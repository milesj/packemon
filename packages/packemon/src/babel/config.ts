import { PluginItem, TransformOptions as ConfigStructure } from '@babel/core';
import { CodeArtifact } from '../CodeArtifact';
import { BROWSER_TARGETS, NATIVE_TARGETS, NODE_SUPPORTED_VERSIONS } from '../constants';
import { FeatureFlags, Format, Platform, Support } from '../types';
import { resolve, resolveFromBabel } from './resolve';

// https://babeljs.io/docs/en/babel-preset-env
export interface PresetEnvOptions {
	browserslistEnv?: string;
	bugfixes?: boolean;
	corejs?: 2 | 3 | { version: 2 | 3; proposals: boolean };
	debug?: boolean;
	exclude?: string[];
	forceAllTransforms?: boolean;
	ignoreBrowserslistConfig?: boolean;
	include?: string[];
	loose?: boolean;
	modules?: 'amd' | 'auto' | 'cjs' | 'commonjs' | 'systemjs' | 'umd' | false;
	shippedProposals?: boolean;
	spec?: boolean;
	targets?: Record<string, string[] | string> | string[] | string;
	useBuiltIns?: 'entry' | 'usage' | false;
}

function getPlatformEnvOptions(
	platform: Platform,
	support: Support,
	format: Format,
): PresetEnvOptions {
	let modules: PresetEnvOptions['modules'] = false;

	if (format === 'umd') {
		modules = 'umd';
	} else if (format === 'cjs' || format === 'lib') {
		modules = 'cjs'; // Babel CommonJS
	}

	switch (platform) {
		case 'browser':
			return {
				modules,
				targets: { browsers: BROWSER_TARGETS[support] },
			};

		case 'native':
			return {
				modules,
				targets: { browsers: NATIVE_TARGETS[support] },
			};

		case 'node':
			return {
				// Async/await has been available since v7
				exclude: [
					'@babel/plugin-transform-regenerator',
					'@babel/plugin-transform-async-to-generator',
				],
				modules,
				targets: {
					node: NODE_SUPPORTED_VERSIONS[support],
				},
			};

		default:
			throw new Error(`Unknown platform "${platform}".`);
	}
}

function getSharedConfig(
	plugins: PluginItem[],
	presets: PluginItem[],
	features: FeatureFlags,
): ConfigStructure {
	return {
		caller: {
			name: 'packemon',
		},
		comments: true,
		parserOpts: {
			sourceType: 'unambiguous',
			strictMode: features.strict,
		},
		plugins,
		presets,
		// Do NOT load root `babel.config.js` as we need full control
		configFile: false,
		// Do load branch `.babelrc.js` files for granular customization
		babelrc: true,
		babelrcRoots: features.workspaces,
	};
}

// The input config should only parse special syntax, not transform and downlevel.
// This applies to all formats within a build target.
export function getBabelInputConfig(
	artifact: CodeArtifact,
	features: FeatureFlags,
): Omit<ConfigStructure, 'exclude' | 'include'> {
	const plugins: PluginItem[] = [];
	const presets: PluginItem[] = [];

	if (features.flow) {
		presets.push([resolve('@babel/preset-flow'), { allowDeclareFields: true }]);
	}

	if (features.typescript) {
		presets.push([resolve('@babel/preset-typescript'), { allowDeclareFields: true }]);

		// When decorators are used, class properties must be loose
		if (features.decorators) {
			plugins.push(
				[resolve('@babel/plugin-proposal-decorators'), { legacy: true }],
				[resolveFromBabel('@babel/plugin-proposal-class-properties'), { loose: true }],
				[resolveFromBabel('@babel/plugin-proposal-private-methods'), { loose: true }],
				[resolveFromBabel('@babel/plugin-proposal-private-property-in-object'), { loose: true }],
			);
		}
	}

	if (features.react) {
		presets.push([
			resolve('@babel/preset-react'),
			{
				development: __DEV__,
				throwIfNamespace: true,
			},
		]);
	}

	return getSharedConfig(plugins, presets, features);
}

// The output config does all the transformation and downleveling through the preset-env.
// This is handled per output since we need to configure based on target + format combinations.
export function getBabelOutputConfig(
	platform: Platform,
	support: Support,
	format: Format,
	features: FeatureFlags,
): ConfigStructure {
	const plugins: PluginItem[] = [];
	const presets: PluginItem[] = [];
	const isESM = format === 'esm' || format === 'mjs';

	// ENVIRONMENT

	const envOptions: PresetEnvOptions = {
		// Prefer spec compliance in development
		spec: __DEV__,
		loose: false,
		// Consumers must polyfill accordingly
		useBuiltIns: false,
		// Transform features accordingly
		bugfixes: true,
		shippedProposals: true,
		// Platform specific
		...getPlatformEnvOptions(platform, support, format),
	};

	presets.push([resolve('@babel/preset-env'), envOptions]);

	// PLUGINS

	if (platform === 'browser' || platform === 'native') {
		// While modern browsers support these features, Node.js <= 12 does not,
		// which results in failing builds trying to parse the syntax.
		// Let's only apply this for the lib format, but allow it for esm.
		if (format === 'lib') {
			plugins.push(
				resolveFromBabel('@babel/plugin-proposal-logical-assignment-operators'),
				resolveFromBabel('@babel/plugin-proposal-nullish-coalescing-operator'),
				resolveFromBabel('@babel/plugin-proposal-optional-chaining'),
			);
		}

		// Both browsers and Node.js support these features outside of legacy targets
		if (support === 'legacy') {
			plugins.push(
				[
					resolve('babel-plugin-transform-async-to-promises'),
					{ inlineHelpers: true, target: 'es5' },
				],
				[
					resolve('@babel/plugin-transform-runtime'),
					{ helpers: false, regenerator: true, useESModules: isESM },
				],
			);
		}
	} else {
		// Use `Object.assign` when available
		plugins.push(
			[resolveFromBabel('@babel/plugin-transform-destructuring'), { useBuiltIns: true }],
			[resolveFromBabel('@babel/plugin-proposal-object-rest-spread'), { useBuiltIns: true }],
		);
	}

	// Support our custom plugins
	if (platform === 'node') {
		plugins.push([resolve('babel-plugin-cjs-esm-interop'), { format: isESM ? 'mjs' : 'cjs' }]);
	}

	plugins.push(
		resolve('babel-plugin-conditional-invariant'),
		resolve('babel-plugin-env-constants'),
	);

	return getSharedConfig(plugins, presets, features);
}
