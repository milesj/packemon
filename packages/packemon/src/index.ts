/**
 * @copyright   2020, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

export * from './Artifact';
export * from './babel/config';
export * from './CodeArtifact';
export * from './commands/Base';
export * from './commands/Build';
export * from './commands/Clean';
export * from './commands/Init';
export * from './commands/Pack';
export * from './commands/Scaffold';
export * from './commands/Validate';
export * from './commands/Watch';
export * from './constants';
export * from './Package';
export * from './PackageValidator';
export * from './Packemon';
export * from './Project';
export * from './rollup/config';
export * from './types';
export * from './TypesArtifact';
