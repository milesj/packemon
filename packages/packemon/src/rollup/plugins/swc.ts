import type { OutputPlugin, Plugin } from 'rollup';
import { type Options, transform } from '@swc/core';

export function swcInput(config: Partial<Options>): Plugin {
	return {
		name: 'packemon-swc-input',

		async transform(code, filename) {
			return transform(code, {
				...config,
				filename,
				sourceMaps: true,
			});
		},
	};
}

export function swcOutput(config: Partial<Options>): OutputPlugin {
	return {
		name: 'packemon-swc-output',

		async renderChunk(code) {
			return transform(code, {
				...config,
				sourceMaps: true,
			});
		},
	};
}
