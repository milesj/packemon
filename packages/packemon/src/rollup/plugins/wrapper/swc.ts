/* eslint-disable no-nested-ternary */

import path from 'path';
import { GetModuleInfo } from 'rollup';
import {
	ArrayPattern,
	ClassDeclaration,
	FunctionDeclaration,
	Identifier,
	Module,
	Node,
	ObjectPattern,
	VariableDeclaration,
} from '@swc/core';
import { ExtractedExports } from './babel';

function extractName(node: Node): string[] | string | undefined {
	switch (node.type) {
		case 'Identifier':
			return (node as Identifier).value;

		// class Foo
		case 'ClassDeclaration':
			return extractName((node as ClassDeclaration).identifier);

		// function foo
		case 'FunctionDeclaration':
			return extractName((node as FunctionDeclaration).identifier);

		// const foo = ...
		case 'VariableDeclaration':
			return (node as VariableDeclaration).declarations
				.map((decl) => extractName(decl.id))
				.filter(Boolean) as string[];

		// const [foo, bar] = ...
		case 'ArrayPattern':
			return (node as ArrayPattern).elements
				.map((el) => (el ? extractName(el) : null))
				.filter(Boolean) as string[];

		// const { foo, bar } = ...
		case 'ObjectPattern':
			return (node as ObjectPattern).properties
				.map((prop) =>
					prop.type === 'KeyValuePatternProperty' ? extractName(prop.value || prop.key) : null,
				)
				.filter(Boolean) as string[];

		default:
			break;
	}

	return undefined;
}

function extractTypeNames(node: Module): string[] {
	const types: string[] = [];

	node.body.forEach((item) => {
		if (
			item.type === 'TsTypeAliasDeclaration' ||
			item.type === 'TsInterfaceDeclaration' ||
			item.type === 'TsEnumDeclaration'
		) {
			types.push(item.id.value);
		}

		if (item.type === 'ImportDeclaration') {
			item.specifiers.forEach((spec) => {
				if (item.typeOnly || ('isTypeOnly' in spec && spec.isTypeOnly)) {
					types.push(spec.local.value);
				}
			});
		}
	});

	return types;
}

export function extractExportsWithSwc(id: string, getModuleInfo: GetModuleInfo): ExtractedExports {
	const info = getModuleInfo(id);

	if (!info || !info.ast) {
		throw new Error(`Cannot get module info for ID: ${id}`);
	}

	const importedFiles = info.importedIds;
	const typeNames = new Set(extractTypeNames(info.ast as unknown as Module));
	const namedExports: string[] = [];
	let defaultExport = false;

	const mapNamed = (name: string[] | string | undefined) => {
		if (Array.isArray(name)) {
			name.forEach((n) => void mapNamed(n));
		} else if (name === 'default') {
			defaultExport = true;
		} else if (name && !typeNames.has(name)) {
			namedExports.push(name);
		}
	};

	console.log(id, info);

	(info.ast as unknown as Module).body.forEach((item) => {
		// export { foo }
		// export foo
		// export * as foo
		if (item.type === 'ExportNamedDeclaration' && !item.typeOnly && item.specifiers.length > 0) {
			mapNamed(
				item.specifiers
					.map((spec) =>
						extractName(
							spec.type === 'ExportNamespaceSpecifier'
								? spec.name
								: spec.type === 'ExportDefaultSpecifier'
								? spec.exported
								: spec.exported ?? spec.orig,
						),
					)
					.filter(Boolean) as string[],
			);
		}

		// export function foo
		// export const foo
		if (item.type === 'ExportDeclaration' && item.declaration) {
			mapNamed(extractName(item.declaration));
		}

		// export default ...
		if (item.type === 'ExportDefaultDeclaration') {
			defaultExport = !!item.decl;
		}

		// export * from ...
		if (item.type === 'ExportAllDeclaration' && item.source) {
			const importId = importedFiles.find((file) =>
				file.startsWith(path.normalize(path.join(path.dirname(id), item.source.value))),
			);

			if (importId) {
				namedExports.push(...extractExportsWithSwc(importId, getModuleInfo).namedExports);
			}
		}
	});

	return { namedExports, defaultExport };
}
