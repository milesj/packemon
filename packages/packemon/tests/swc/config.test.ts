/* eslint-disable no-param-reassign */

import { getSwcInputConfig, getSwcOutputConfig } from '../../src/swc/config';
import { Format, Platform, Support } from '../../src/types';

const SUPPORTS: Support[] = ['legacy', 'stable', 'current', 'experimental'];

describe('getSwcInputConfig()', () => {
	const bundleArtifact: any = {
		package: { hasDependency: () => false },
	};

	it('includes no parser/transformer options by default', () => {
		expect(getSwcInputConfig(bundleArtifact, {})).toMatchSnapshot();
	});

	it('includes react transformer if `react` feature flag is true', () => {
		expect(
			getSwcInputConfig(bundleArtifact, { react: 'automatic' }).jsc?.transform,
		).toMatchSnapshot();
	});

	it('includes typescript parser if `typescript` feature flag is true', () => {
		expect(getSwcInputConfig(bundleArtifact, { typescript: true }).jsc?.parser).toMatchSnapshot();
	});

	it('includes typescript decorators if `typescript` and `decorators` feature flag is true', () => {
		expect(
			getSwcInputConfig(bundleArtifact, { decorators: true, typescript: true }),
		).toMatchSnapshot();
	});

	it('doesnt include typescript decorators if `typescript` feature flag is false', () => {
		expect(
			getSwcInputConfig(bundleArtifact, { decorators: true, typescript: false }),
		).toMatchSnapshot();
	});

	it('supports private properties with decorators if dep exists', () => {
		const spy = jest.spyOn(bundleArtifact.package, 'hasDependency').mockImplementation(() => true);

		expect(
			getSwcInputConfig(bundleArtifact, { decorators: true, typescript: true }),
		).toMatchSnapshot();

		spy.mockRestore();
	});

	it('can mutate config', () => {
		expect(
			getSwcInputConfig(
				bundleArtifact,
				{},
				{
					swcInput(config) {
						config.rootMode = 'upward';
						config.cwd = 'INPUT';
					},
					swcOutput(config) {
						config.cwd = 'OUTPUT';
					},
				},
			),
		).toMatchSnapshot();
	});
});

function renderPresetEnv(platform: Platform, format: Format, support: Support) {
	// eslint-disable-next-line jest/require-top-level-describe
	test(`handles preset-env: ${platform} + ${format} + ${support}`, () => {
		expect(getSwcOutputConfig(platform, support, format, {})?.env).toMatchSnapshot();
	});
}

describe('getSwcOutputConfig()', () => {
	SUPPORTS.forEach((support) => {
		renderPresetEnv('native', 'lib', support);

		(['lib', 'cjs', 'mjs'] as const).forEach((format) => {
			renderPresetEnv('node', format, support);
		});

		(['lib', 'esm', 'umd'] as const).forEach((format) => {
			renderPresetEnv('browser', format, support);
		});
	});

	it('errors for invalid platform', () => {
		expect(() =>
			// @ts-expect-error Unknown platform
			getSwcOutputConfig('unknown', 'stable', 'lib', {}),
		).toThrow('Unknown platform "unknown".');
	});

	it('transforms async/await to promises when `browser` or `native`', () => {
		expect(getSwcOutputConfig('browser', 'stable', 'lib', {})).toMatchSnapshot();

		expect(getSwcOutputConfig('native', 'experimental', 'lib', {})).toMatchSnapshot();
	});

	it('uses built-in destructuring and object spread when `current` or `experimental`', () => {
		expect(getSwcOutputConfig('node', 'current', 'lib', {})).toMatchSnapshot();

		expect(getSwcOutputConfig('node', 'experimental', 'lib', {})).toMatchSnapshot();
	});

	it('can mutate config', () => {
		expect(
			getSwcOutputConfig(
				'browser',
				'stable',
				'lib',
				{},
				{
					swcInput(config) {
						config.cwd = 'INPUT';
					},
					swcOutput(config) {
						config.rootMode = 'upward';
						config.cwd = 'OUTPUT';
					},
				},
			),
		).toMatchSnapshot();
	});

	it('passes build params to config', () => {
		const spy = jest.fn();

		getSwcOutputConfig(
			'browser',
			'stable',
			'lib',
			{},
			{
				swcOutput: spy,
			},
		);

		expect(spy).toHaveBeenCalledWith(expect.any(Object), {
			features: {},
			format: 'lib',
			platform: 'browser',
			support: 'stable',
		});
	});
});
