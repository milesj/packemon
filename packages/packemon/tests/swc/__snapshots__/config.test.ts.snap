// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getSwcInputConfig() can mutate config 1`] = `
Object {
  "caller": Object {
    "name": "packemon",
  },
  "configFile": false,
  "cwd": "INPUT",
  "jsc": Object {
    "externalHelpers": false,
    "keepClassNames": true,
    "loose": false,
    "parser": Object {
      "dynamicImport": true,
      "jsx": false,
      "syntax": "ecmascript",
    },
    "preserveAllComments": true,
    "target": "es2022",
    "transform": Object {
      "optimizer": undefined,
    },
  },
  "rootMode": "upward",
  "swcrc": false,
}
`;

exports[`getSwcInputConfig() doesnt include typescript decorators if \`typescript\` feature flag is false 1`] = `
Object {
  "caller": Object {
    "name": "packemon",
  },
  "configFile": false,
  "jsc": Object {
    "externalHelpers": false,
    "keepClassNames": true,
    "loose": false,
    "parser": Object {
      "dynamicImport": true,
      "jsx": false,
      "syntax": "ecmascript",
    },
    "preserveAllComments": true,
    "target": "es2022",
    "transform": Object {
      "optimizer": undefined,
    },
  },
  "swcrc": false,
}
`;

exports[`getSwcInputConfig() includes no parser/transformer options by default 1`] = `
Object {
  "caller": Object {
    "name": "packemon",
  },
  "configFile": false,
  "jsc": Object {
    "externalHelpers": false,
    "keepClassNames": true,
    "loose": false,
    "parser": Object {
      "dynamicImport": true,
      "jsx": false,
      "syntax": "ecmascript",
    },
    "preserveAllComments": true,
    "target": "es2022",
    "transform": Object {
      "optimizer": undefined,
    },
  },
  "swcrc": false,
}
`;

exports[`getSwcInputConfig() includes react transformer if \`react\` feature flag is true 1`] = `
Object {
  "optimizer": undefined,
  "react": Object {
    "development": true,
    "runtime": "automatic",
    "throwIfNamespace": true,
  },
}
`;

exports[`getSwcInputConfig() includes typescript decorators if \`typescript\` and \`decorators\` feature flag is true 1`] = `
Object {
  "caller": Object {
    "name": "packemon",
  },
  "configFile": false,
  "jsc": Object {
    "externalHelpers": false,
    "keepClassNames": true,
    "loose": false,
    "parser": Object {
      "decorators": true,
      "dynamicImport": true,
      "syntax": "typescript",
      "tsx": false,
    },
    "preserveAllComments": true,
    "target": "es2022",
    "transform": Object {
      "legacyDecorator": true,
      "optimizer": undefined,
    },
  },
  "swcrc": false,
}
`;

exports[`getSwcInputConfig() includes typescript parser if \`typescript\` feature flag is true 1`] = `
Object {
  "dynamicImport": true,
  "syntax": "typescript",
  "tsx": false,
}
`;

exports[`getSwcInputConfig() supports private properties with decorators if dep exists 1`] = `
Object {
  "caller": Object {
    "name": "packemon",
  },
  "configFile": false,
  "jsc": Object {
    "externalHelpers": false,
    "keepClassNames": true,
    "loose": false,
    "parser": Object {
      "decorators": true,
      "dynamicImport": true,
      "syntax": "typescript",
      "tsx": false,
    },
    "preserveAllComments": true,
    "target": "es2022",
    "transform": Object {
      "legacyDecorator": true,
      "optimizer": undefined,
    },
  },
  "swcrc": false,
}
`;

exports[`getSwcOutputConfig() can mutate config 1`] = `
Object {
  "caller": Object {
    "name": "packemon",
  },
  "configFile": false,
  "cwd": "OUTPUT",
  "env": Object {
    "bugfixes": true,
    "include": Array [
      "proposal-logical-assignment-operators",
      "proposal-nullish-coalescing-operator",
      "proposal-optional-chaining",
    ],
    "loose": false,
    "mode": undefined,
    "shippedProposals": true,
    "targets": "defaults, not IE 11",
  },
  "jsc": Object {
    "keepClassNames": true,
    "parser": Object {
      "syntax": "ecmascript",
    },
    "preserveAllComments": true,
    "target": "es2018",
    "transform": Object {
      "optimizer": Object {
        "globals": Object {
          "vars": Object {
            "__DEV__": "process.env.NODE_ENV !== 'production'",
            "__PROD__": "process.env.NODE_ENV === 'production'",
            "__TEST__": "process.env.NODE_ENV === 'test'",
          },
        },
      },
    },
  },
  "module": Object {
    "ignoreDynamic": true,
    "type": "commonjs",
  },
  "rootMode": "upward",
  "swcrc": false,
}
`;

exports[`getSwcOutputConfig() handles preset-env: browser + esm + current 1`] = `
Object {
  "bugfixes": true,
  "loose": false,
  "mode": undefined,
  "shippedProposals": true,
  "targets": "> 1%, not dead",
}
`;

exports[`getSwcOutputConfig() handles preset-env: browser + esm + experimental 1`] = `
Object {
  "bugfixes": true,
  "loose": false,
  "mode": undefined,
  "shippedProposals": true,
  "targets": "last 2 chrome versions, last 2 firefox versions",
}
`;

exports[`getSwcOutputConfig() handles preset-env: browser + esm + legacy 1`] = `
Object {
  "bugfixes": true,
  "loose": false,
  "mode": undefined,
  "shippedProposals": true,
  "targets": "IE 11",
}
`;

exports[`getSwcOutputConfig() handles preset-env: browser + esm + stable 1`] = `
Object {
  "bugfixes": true,
  "loose": false,
  "mode": undefined,
  "shippedProposals": true,
  "targets": "defaults, not IE 11",
}
`;

exports[`getSwcOutputConfig() handles preset-env: browser + lib + current 1`] = `
Object {
  "bugfixes": true,
  "include": Array [
    "proposal-logical-assignment-operators",
    "proposal-nullish-coalescing-operator",
    "proposal-optional-chaining",
  ],
  "loose": false,
  "mode": undefined,
  "shippedProposals": true,
  "targets": "> 1%, not dead",
}
`;

exports[`getSwcOutputConfig() handles preset-env: browser + lib + experimental 1`] = `
Object {
  "bugfixes": true,
  "include": Array [
    "proposal-logical-assignment-operators",
    "proposal-nullish-coalescing-operator",
    "proposal-optional-chaining",
  ],
  "loose": false,
  "mode": undefined,
  "shippedProposals": true,
  "targets": "last 2 chrome versions, last 2 firefox versions",
}
`;

exports[`getSwcOutputConfig() handles preset-env: browser + lib + legacy 1`] = `
Object {
  "bugfixes": true,
  "include": Array [
    "proposal-logical-assignment-operators",
    "proposal-nullish-coalescing-operator",
    "proposal-optional-chaining",
  ],
  "loose": false,
  "mode": undefined,
  "shippedProposals": true,
  "targets": "IE 11",
}
`;

exports[`getSwcOutputConfig() handles preset-env: browser + lib + stable 1`] = `
Object {
  "bugfixes": true,
  "include": Array [
    "proposal-logical-assignment-operators",
    "proposal-nullish-coalescing-operator",
    "proposal-optional-chaining",
  ],
  "loose": false,
  "mode": undefined,
  "shippedProposals": true,
  "targets": "defaults, not IE 11",
}
`;

exports[`getSwcOutputConfig() handles preset-env: browser + umd + current 1`] = `
Object {
  "bugfixes": true,
  "loose": false,
  "mode": undefined,
  "shippedProposals": true,
  "targets": "> 1%, not dead",
}
`;

exports[`getSwcOutputConfig() handles preset-env: browser + umd + experimental 1`] = `
Object {
  "bugfixes": true,
  "loose": false,
  "mode": undefined,
  "shippedProposals": true,
  "targets": "last 2 chrome versions, last 2 firefox versions",
}
`;

exports[`getSwcOutputConfig() handles preset-env: browser + umd + legacy 1`] = `
Object {
  "bugfixes": true,
  "loose": false,
  "mode": undefined,
  "shippedProposals": true,
  "targets": "IE 11",
}
`;

exports[`getSwcOutputConfig() handles preset-env: browser + umd + stable 1`] = `
Object {
  "bugfixes": true,
  "loose": false,
  "mode": undefined,
  "shippedProposals": true,
  "targets": "defaults, not IE 11",
}
`;

exports[`getSwcOutputConfig() handles preset-env: native + lib + current 1`] = `
Object {
  "bugfixes": true,
  "include": Array [
    "proposal-logical-assignment-operators",
    "proposal-nullish-coalescing-operator",
    "proposal-optional-chaining",
  ],
  "loose": false,
  "mode": undefined,
  "shippedProposals": true,
  "targets": "iOS 12",
}
`;

exports[`getSwcOutputConfig() handles preset-env: native + lib + experimental 1`] = `
Object {
  "bugfixes": true,
  "include": Array [
    "proposal-logical-assignment-operators",
    "proposal-nullish-coalescing-operator",
    "proposal-optional-chaining",
  ],
  "loose": false,
  "mode": undefined,
  "shippedProposals": true,
  "targets": "iOS 14",
}
`;

exports[`getSwcOutputConfig() handles preset-env: native + lib + legacy 1`] = `
Object {
  "bugfixes": true,
  "include": Array [
    "proposal-logical-assignment-operators",
    "proposal-nullish-coalescing-operator",
    "proposal-optional-chaining",
  ],
  "loose": false,
  "mode": undefined,
  "shippedProposals": true,
  "targets": "iOS 8",
}
`;

exports[`getSwcOutputConfig() handles preset-env: native + lib + stable 1`] = `
Object {
  "bugfixes": true,
  "include": Array [
    "proposal-logical-assignment-operators",
    "proposal-nullish-coalescing-operator",
    "proposal-optional-chaining",
  ],
  "loose": false,
  "mode": undefined,
  "shippedProposals": true,
  "targets": "iOS 10",
}
`;

exports[`getSwcOutputConfig() handles preset-env: node + cjs + current 1`] = `
Object {
  "bugfixes": true,
  "loose": false,
  "mode": undefined,
  "shippedProposals": true,
  "targets": Object {
    "node": "14.16.0",
  },
}
`;

exports[`getSwcOutputConfig() handles preset-env: node + cjs + experimental 1`] = `
Object {
  "bugfixes": true,
  "loose": false,
  "mode": undefined,
  "shippedProposals": true,
  "targets": Object {
    "node": "16.0.0",
  },
}
`;

exports[`getSwcOutputConfig() handles preset-env: node + cjs + legacy 1`] = `
Object {
  "bugfixes": true,
  "loose": false,
  "mode": undefined,
  "shippedProposals": true,
  "targets": Object {
    "node": "10.3.0",
  },
}
`;

exports[`getSwcOutputConfig() handles preset-env: node + cjs + stable 1`] = `
Object {
  "bugfixes": true,
  "loose": false,
  "mode": undefined,
  "shippedProposals": true,
  "targets": Object {
    "node": "12.17.0",
  },
}
`;

exports[`getSwcOutputConfig() handles preset-env: node + lib + current 1`] = `
Object {
  "bugfixes": true,
  "loose": false,
  "mode": undefined,
  "shippedProposals": true,
  "targets": Object {
    "node": "14.16.0",
  },
}
`;

exports[`getSwcOutputConfig() handles preset-env: node + lib + experimental 1`] = `
Object {
  "bugfixes": true,
  "loose": false,
  "mode": undefined,
  "shippedProposals": true,
  "targets": Object {
    "node": "16.0.0",
  },
}
`;

exports[`getSwcOutputConfig() handles preset-env: node + lib + legacy 1`] = `
Object {
  "bugfixes": true,
  "loose": false,
  "mode": undefined,
  "shippedProposals": true,
  "targets": Object {
    "node": "10.3.0",
  },
}
`;

exports[`getSwcOutputConfig() handles preset-env: node + lib + stable 1`] = `
Object {
  "bugfixes": true,
  "loose": false,
  "mode": undefined,
  "shippedProposals": true,
  "targets": Object {
    "node": "12.17.0",
  },
}
`;

exports[`getSwcOutputConfig() handles preset-env: node + mjs + current 1`] = `
Object {
  "bugfixes": true,
  "loose": false,
  "mode": undefined,
  "shippedProposals": true,
  "targets": Object {
    "node": "14.16.0",
  },
}
`;

exports[`getSwcOutputConfig() handles preset-env: node + mjs + experimental 1`] = `
Object {
  "bugfixes": true,
  "loose": false,
  "mode": undefined,
  "shippedProposals": true,
  "targets": Object {
    "node": "16.0.0",
  },
}
`;

exports[`getSwcOutputConfig() handles preset-env: node + mjs + legacy 1`] = `
Object {
  "bugfixes": true,
  "loose": false,
  "mode": undefined,
  "shippedProposals": true,
  "targets": Object {
    "node": "10.3.0",
  },
}
`;

exports[`getSwcOutputConfig() handles preset-env: node + mjs + stable 1`] = `
Object {
  "bugfixes": true,
  "loose": false,
  "mode": undefined,
  "shippedProposals": true,
  "targets": Object {
    "node": "12.17.0",
  },
}
`;

exports[`getSwcOutputConfig() transforms async/await to promises when \`browser\` or \`native\` 1`] = `
Object {
  "caller": Object {
    "name": "packemon",
  },
  "configFile": false,
  "env": Object {
    "bugfixes": true,
    "include": Array [
      "proposal-logical-assignment-operators",
      "proposal-nullish-coalescing-operator",
      "proposal-optional-chaining",
    ],
    "loose": false,
    "mode": undefined,
    "shippedProposals": true,
    "targets": "defaults, not IE 11",
  },
  "jsc": Object {
    "keepClassNames": true,
    "parser": Object {
      "syntax": "ecmascript",
    },
    "preserveAllComments": true,
    "target": "es2018",
    "transform": Object {
      "optimizer": Object {
        "globals": Object {
          "vars": Object {
            "__DEV__": "process.env.NODE_ENV !== 'production'",
            "__PROD__": "process.env.NODE_ENV === 'production'",
            "__TEST__": "process.env.NODE_ENV === 'test'",
          },
        },
      },
    },
  },
  "module": Object {
    "ignoreDynamic": true,
    "type": "commonjs",
  },
  "swcrc": false,
}
`;

exports[`getSwcOutputConfig() transforms async/await to promises when \`browser\` or \`native\` 2`] = `
Object {
  "caller": Object {
    "name": "packemon",
  },
  "configFile": false,
  "env": Object {
    "bugfixes": true,
    "include": Array [
      "proposal-logical-assignment-operators",
      "proposal-nullish-coalescing-operator",
      "proposal-optional-chaining",
    ],
    "loose": false,
    "mode": undefined,
    "shippedProposals": true,
    "targets": "iOS 14",
  },
  "jsc": Object {
    "keepClassNames": true,
    "parser": Object {
      "syntax": "ecmascript",
    },
    "preserveAllComments": true,
    "target": "es2022",
    "transform": Object {
      "optimizer": Object {
        "globals": Object {
          "vars": Object {
            "__DEV__": "process.env.NODE_ENV !== 'production'",
            "__PROD__": "process.env.NODE_ENV === 'production'",
            "__TEST__": "process.env.NODE_ENV === 'test'",
          },
        },
      },
    },
  },
  "module": Object {
    "ignoreDynamic": true,
    "type": "commonjs",
  },
  "swcrc": false,
}
`;

exports[`getSwcOutputConfig() uses built-in destructuring and object spread when \`current\` or \`experimental\` 1`] = `
Object {
  "caller": Object {
    "name": "packemon",
  },
  "configFile": false,
  "env": Object {
    "bugfixes": true,
    "loose": false,
    "mode": undefined,
    "shippedProposals": true,
    "targets": Object {
      "node": "14.16.0",
    },
  },
  "jsc": Object {
    "keepClassNames": true,
    "parser": Object {
      "syntax": "ecmascript",
    },
    "preserveAllComments": true,
    "target": "es2020",
    "transform": Object {
      "optimizer": Object {
        "globals": Object {
          "vars": Object {
            "__DEV__": "process.env.NODE_ENV !== 'production'",
            "__PROD__": "process.env.NODE_ENV === 'production'",
            "__TEST__": "process.env.NODE_ENV === 'test'",
          },
        },
      },
    },
  },
  "module": Object {
    "ignoreDynamic": true,
    "type": "commonjs",
  },
  "swcrc": false,
}
`;

exports[`getSwcOutputConfig() uses built-in destructuring and object spread when \`current\` or \`experimental\` 2`] = `
Object {
  "caller": Object {
    "name": "packemon",
  },
  "configFile": false,
  "env": Object {
    "bugfixes": true,
    "loose": false,
    "mode": undefined,
    "shippedProposals": true,
    "targets": Object {
      "node": "16.0.0",
    },
  },
  "jsc": Object {
    "keepClassNames": true,
    "parser": Object {
      "syntax": "ecmascript",
    },
    "preserveAllComments": true,
    "target": "es2022",
    "transform": Object {
      "optimizer": Object {
        "globals": Object {
          "vars": Object {
            "__DEV__": "process.env.NODE_ENV !== 'production'",
            "__PROD__": "process.env.NODE_ENV === 'production'",
            "__TEST__": "process.env.NODE_ENV === 'test'",
          },
        },
      },
    },
  },
  "module": Object {
    "ignoreDynamic": true,
    "type": "commonjs",
  },
  "swcrc": false,
}
`;
